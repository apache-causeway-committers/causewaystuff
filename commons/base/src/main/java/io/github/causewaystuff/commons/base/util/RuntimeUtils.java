/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.commons.base.util;

import java.util.concurrent.atomic.AtomicReference;

import org.apache.causeway.applib.services.factory.FactoryService;
import org.apache.causeway.applib.services.inject.ServiceInjector;
import org.apache.causeway.applib.services.repository.RepositoryService;
import org.apache.causeway.applib.services.wrapper.WrapperFactory;
import org.apache.causeway.commons.internal.context._Context;
import org.apache.causeway.commons.internal.ioc._IocContainer;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class RuntimeUtils {

    @Getter(lazy=true)
    private final FactoryService factoryService = getIocContainer().get(FactoryService.class).orElseThrow();

    @Getter(lazy=true)
    private final ServiceInjector serviceInjector = getIocContainer().get(ServiceInjector.class).orElseThrow();

    @Getter(lazy=true)
    private final RepositoryService repositoryService = getIocContainer().get(RepositoryService.class).orElseThrow();

    @Getter(lazy=true)
    private final WrapperFactory wrapperFactory = getIocContainer().get(WrapperFactory.class).orElseThrow();

    /**
     * Invalidates cached values, potentially useful in case the underlying application context was refreshed.
     */
    public void invalidate() {
        ((AtomicReference<Object>)factoryService).set(null);
        ((AtomicReference<Object>)serviceInjector).set(null);
        ((AtomicReference<Object>)repositoryService).set(null);
        ((AtomicReference<Object>)wrapperFactory).set(null);
        ((AtomicReference<Object>)iocContainer).set(null);
    }

    // -- HELPER

    @Getter(lazy=true, value = AccessLevel.PRIVATE)
    private final _IocContainer iocContainer = iocContainer();

    //TODO provide friendly error messages (requires a valid/bootstrapped Spring context; otherwise will fail)
    @SneakyThrows
    private _IocContainer iocContainer() {
        var envClass = _Context
            .loadClass("org.apache.causeway.core.config.environment.CausewaySystemEnvironment");
        var env = _Context.getElseFail(envClass);
        final _IocContainer iocContainer = (_IocContainer) envClass.getMethod("getIocContainer")
                .invoke(env);
        return iocContainer;
    }

}
