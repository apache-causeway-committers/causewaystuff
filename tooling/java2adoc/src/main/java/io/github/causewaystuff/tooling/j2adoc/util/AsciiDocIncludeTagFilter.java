/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.tooling.j2adoc.util;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.stream.Collectors;

import org.apache.causeway.commons.io.TextUtils;

public final class AsciiDocIncludeTagFilter {

    public static String read(final File source) {
        return TextUtils.readLinesFromFile(source, StandardCharsets.UTF_8).stream()
        //.filter(line->!containsIncludeTag(line))
        .filter(line->!isAllLineComment(line))
        .map(AsciiDocIncludeTagFilter::removeFootNoteReference)
        .collect(Collectors.joining("\n"));
    }

    public static void removeAdocExampleTags(final File source) {
        var fixedLines = TextUtils.readLinesFromFile(source, StandardCharsets.UTF_8)
        .filter(line->!isIncludeTagComment(line))
        .map(AsciiDocIncludeTagFilter::removeFootNoteReference);

        TextUtils.writeLinesToFile(fixedLines, source, StandardCharsets.UTF_8);
    }

    // -- HELPER

    private static boolean isIncludeTagComment(String line) {
        line = line.trim();
        if(!line.startsWith("//")) {
            return false;
        }
        return line.contains(" tag::")
                || line.contains(" end::");
    }

    private static boolean isAllLineComment(final String line) {
        return line.trim().startsWith("//");
    }

    private static String removeFootNoteReference(final String line) {
        if(!line.contains("// <")) {
            return line;
        }
        return line.replace("// <.>", "")
                .replace("// <1>", "")
                .replace("// <2>", "")
                .replace("// <3>", "")
                .replace("// <4>", "")
                .replace("// <5>", "")
                .replace("// <6>", "")
                .replace("// <7>", "")
                .replace("// <8>", "")
                .replace("// <9>", "")
                .replace("// <10>", "")
                .replace("// <11>", "")
                .replace("// <12>", "")
                .replace("// <13>", "")
                .replace("// <14>", "")
                .replace("// <15>", "")
                .replace("// <16>", "")
                .stripTrailing();
    }

}
