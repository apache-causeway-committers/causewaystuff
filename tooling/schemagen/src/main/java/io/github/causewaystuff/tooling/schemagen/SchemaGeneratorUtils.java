/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.tooling.schemagen;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.github.victools.jsonschema.generator.Option;
import com.github.victools.jsonschema.generator.OptionPreset;
import com.github.victools.jsonschema.generator.SchemaGenerator;
import com.github.victools.jsonschema.generator.SchemaGeneratorConfig;
import com.github.victools.jsonschema.generator.SchemaGeneratorConfigBuilder;
import com.github.victools.jsonschema.generator.SchemaVersion;
import com.github.victools.jsonschema.generator.impl.TypeContextFactory;
import com.github.victools.jsonschema.generator.impl.module.MethodExclusionModule;

import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class SchemaGeneratorUtils {

    @SneakyThrows
    public SchemaGenerator schemaGeneratorDefault() {
        var config = defaultConfig();
        var typeContext = TypeContextFactory.createDefaultTypeContext(config);
        return new SchemaGenerator(config, typeContext);
    }
    
    @SneakyThrows
    public SchemaGenerator schemaGeneratorWithAnnotationTypeSupport() {
        var config = defaultConfig();
        var typeContext = TypeContextFactory.createDefaultTypeContext(config);
        return new SchemaGenerator(config, SchemaGeneratorPatcher.patch(typeContext));
    }
    
    @SneakyThrows
    public String prettyPrint(JsonNode jsonSchema) {
        var objectMapper = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT); // pretty-printing
        var prettyJsonString = objectMapper
            .writerWithDefaultPrettyPrinter()
            .writeValueAsString(jsonSchema);
        return prettyJsonString;
    }
    
    // -- HELPER
    
    private SchemaGeneratorConfig defaultConfig() {
        var configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2020_12, OptionPreset.PLAIN_JSON);
        var config = configBuilder
                .with(new MethodExclusionModule(methodScope->
                        methodScope.getRawMember().getName().equals("toString")
                        || methodScope.getRawMember().getParameterCount()>0))
                .with(Option.NONSTATIC_NONVOID_NONGETTER_METHODS)
                .with(Option.EXTRA_OPEN_API_FORMAT_VALUES)
                .without(Option.FLATTENED_ENUMS_FROM_TOSTRING)
                .build();
        return config;
    }

}
