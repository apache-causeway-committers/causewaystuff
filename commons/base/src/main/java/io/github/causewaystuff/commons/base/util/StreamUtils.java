/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.commons.base.util;

import java.util.Iterator;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class StreamUtils {

    public <T> Stream<T> toStream(final Iterator<T> sourceIterator) {
        return toStream(sourceIterator, false);
    }

    public <T> Stream<T> toStream(final Iterator<T> sourceIterator, final boolean parallel) {
        Iterable<T> iterable = () -> sourceIterator;
        return StreamSupport.stream(iterable.spliterator(), parallel);
    }

    @FunctionalInterface
    public static interface Collector<T> {
        void collect(T t);
        default void terminate() {
            collect(null); // terminal
        }
    }

    /**
     * Creates a {@link Stream} from a producer that supports the visitor pattern.
     * <p>
     * A null element from the producer terminates the sequence.
     * <p>
     * Limitation: an element once produced must not be modified by the producer later,
     *      as 2 threads are racing for access to the element
     * <p>
     * Usage:
     * <pre>
     * StreamUtils.toStream(collector->{
     *      someProducer.visit(collector::collect);
     * });
     * </pre>
     * @see Collector
     */
    @SneakyThrows
    public <T> Stream<T> toStream(final Consumer<Collector<T>> consumer) {
        final var collector = new BlockingCollector<T>(
                new LinkedBlockingQueue<>(1),
                new AtomicBoolean(false));
        var producer = Thread.startVirtualThread(()->{
            consumer.accept(collector);
            collector.terminate();
        });

        return Stream.generate(collector::take)
                .takeWhile(t->{
                    var terminated = t==null;
                    if(terminated) {
                        try {
                            producer.join();
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                    return !terminated;
                });
    }

    // -- HELPER

    private record BlockingCollector<T>(
            LinkedBlockingQueue<T> q,
            AtomicBoolean terminated)
    implements Collector<T> {
        @SneakyThrows
        @Override
        public void collect(final T t) {
            if(t==null) {
                terminated.set(true);
                return;
            }
            q.put(t);
        }
        /**
         * Try to poll until q is empty or terminated.
         */
        @SneakyThrows
        public T take() {
            while(hasPotentiallyMore()) {
                var t = q.poll(10, TimeUnit.MILLISECONDS);
                if(t!=null) {
                    return t;
                }
            }
            return null;
        }

        // -- HELPER

        private boolean hasPotentiallyMore() {
            if(terminated.get()) {
                return !q.isEmpty();
            }
            return true;
        }

    }

}
