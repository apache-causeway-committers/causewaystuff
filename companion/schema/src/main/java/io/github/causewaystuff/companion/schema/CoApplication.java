/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.companion.schema;

import java.util.List;

import org.apache.causeway.commons.io.YamlUtils;

public record CoApplication(
    String groupId,
    String artifactId,
    String version,
    String name, 
    String description,
    String packageName,
    Persistence persistence,
    List<CoModule> modules) {

    public static CoApplication fromYaml(String yaml) {
        return YamlUtils.tryRead(CoApplication.class, yaml)
            .valueAsNonNullElseFail();
    }
    
    public String toYaml() {
        return YamlUtils.toStringUtf8(this);
    }
}
