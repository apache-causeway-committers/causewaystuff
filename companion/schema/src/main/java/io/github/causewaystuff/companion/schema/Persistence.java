/*
v *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.companion.schema;

import org.springframework.util.StringUtils;

public enum Persistence {
    NONE,
    JDBC,
    JPA,
    JDO;
    
    public boolean isNone() { return this==NONE; }
    public boolean isJdbc() { return this==JDBC; }
    public boolean isJpa() { return this==JPA; }
    public boolean isJdo() { return this==JDO; }
    
    public static Persistence parse(final String value) {
        if(!StringUtils.hasLength(value)) return NONE;
        return switch (value.toLowerCase()) {
            case "jdbc" -> JDBC;
            case "jpa" -> JPA;
            case "jdo" -> JDO;
            default -> NONE;
        };
    }
}
