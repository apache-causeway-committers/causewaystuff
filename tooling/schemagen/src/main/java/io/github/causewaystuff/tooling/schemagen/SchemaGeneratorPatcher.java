/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.tooling.schemagen;

import java.lang.reflect.Type;
import java.util.List;

import com.fasterxml.classmate.ResolvedType;
import com.fasterxml.classmate.TypeBindings;
import com.fasterxml.classmate.TypeResolver;
import com.fasterxml.classmate.members.RawMethod;
import com.fasterxml.classmate.types.ResolvedObjectType;
import com.github.victools.jsonschema.generator.TypeContext;

import org.apache.causeway.commons.collections.Can;
import org.apache.causeway.commons.internal.reflection._Reflect;

import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
class SchemaGeneratorPatcher {

    /**
     * Patch the type resolver with our own, to support annotation types.
     * @return
     */
    @SneakyThrows
    TypeContext patch(final TypeContext typeContext) {
        _Reflect.setFieldOn(TypeContext.class.getDeclaredField("typeResolver"), typeContext, new TypeResolver2());
        return typeContext;
    }

    // -- HELPER

    private static class ResolvedType2 extends ResolvedObjectType {
        public ResolvedType2(
                final Class<?> erased,
                final TypeBindings bindings,
                final ResolvedType superClass,
                final List<ResolvedType> interfaces,
                final Can<RawMethod> memberMethods) {
            super(erased, bindings, superClass, interfaces);
            this._memberMethods = memberMethods.toArray(new RawMethod[0]);
        }
        @Override
        public synchronized List<RawMethod> getMemberMethods() {
            return super.getMemberMethods();
        }
    }

    private static class TypeResolver2 extends TypeResolver {
        private static final long serialVersionUID = 1L;
        @Override
        public ResolvedType resolve(final Type type, final Type... typeParameters) {
            var t = super.resolve(type, typeParameters);
            // intercept annotation types
            if(t.getErasedType().isAnnotation()) {
                var methods = Can.ofArray(t.getErasedType().getMethods())
                        .stream()
                        .limit(16)
                        .map(method->new RawMethod(t, method))
                        .collect(Can.toCan());
                return new ResolvedType2(t.getErasedType(), t.getTypeBindings(), null, null, methods);
            }
            return t;
        }
    }

}
