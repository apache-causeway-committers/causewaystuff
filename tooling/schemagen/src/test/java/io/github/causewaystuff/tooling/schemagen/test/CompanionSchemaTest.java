/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.tooling.schemagen.test;

import java.io.IOException;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import org.apache.causeway.testing.unittestsupport.applib.util.ApprovalUtils;

import io.github.causewaystuff.companion.schema.CoApplication;
import io.github.causewaystuff.tooling.schemagen.SchemaGeneratorUtils;

class CompanionSchemaTest {

    @Test
    @UseReporter(DiffReporter.class)
    void coApplication() throws IOException  {
        var generator = SchemaGeneratorUtils.schemaGeneratorWithRecordTypeSupport();
        var jsonSchema = generator.generateSchema(CoApplication.class);
        var json = SchemaGeneratorUtils.prettyPrint(jsonSchema);
        
        Approvals.verify(json, ApprovalUtils.ignoreLineEndings()
            .forFile().withExtension(".json"));
    }

}
