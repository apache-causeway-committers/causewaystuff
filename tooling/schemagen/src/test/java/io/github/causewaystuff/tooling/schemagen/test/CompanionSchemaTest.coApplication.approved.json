{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "type" : "object",
  "properties" : {
    "artifactId" : {
      "type" : "string"
    },
    "description" : {
      "type" : "string"
    },
    "groupId" : {
      "type" : "string"
    },
    "modules" : {
      "type" : "array",
      "items" : {
        "type" : "object",
        "properties" : {
          "description" : {
            "type" : "string"
          },
          "entities" : {
            "type" : "array",
            "items" : {
              "type" : "object",
              "properties" : {
                "description" : {
                  "type" : "string"
                },
                "fields" : {
                  "type" : "array",
                  "items" : {
                    "type" : "object",
                    "properties" : {
                      "description" : {
                        "type" : "string"
                      },
                      "flags" : {
                        "type" : "array",
                        "items" : {
                          "type" : "string",
                          "enum" : [ "PRIMARY_KEY", "SECONDARY_KEY" ]
                        }
                      },
                      "name" : {
                        "type" : "string"
                      },
                      "typeFqn" : {
                        "type" : "string"
                      }
                    }
                  }
                },
                "name" : {
                  "type" : "string"
                },
                "namespace" : {
                  "type" : "string"
                }
              }
            }
          },
          "id" : {
            "type" : "string"
          },
          "name" : {
            "type" : "string"
          }
        }
      }
    },
    "name" : {
      "type" : "string"
    },
    "packageName" : {
      "type" : "string"
    },
    "persistence" : {
      "type" : "string",
      "enum" : [ "NONE", "JPA", "JDO" ]
    },
    "version" : {
      "type" : "string"
    }
  }
}